docs/dev/libs/cep.md

```diff
@@ -98,48 +128,105 @@ val result: DataStream[Alert] = patternStream.select(createAlert(_))
+can be seen as a graph of such states, where transition from one state to the next happens based on user-specified
+*conditions*, e.g. `event.getName().equals("start")`. A *match* is a sequence of input events which visit all 
+states of the complex pattern graph, through a sequence of valid state transitions.
+
+<span class="label label-danger">Attention</span> Each state must have a unique name to identify the matched 
+events later on. 
+
+<span class="label label-danger">Attention</span> State names **CANNOT** contain the character `:`.
+
+In the remainder, we start by describing how to define [States](#states), before describing how you can 
+combine individual states into [Complex Patterns](#combining-states).
+
+### Individual States
+
+A **State** can be either a *singleton* state, or a *looping* one. Singleton states accept a single event, 
+while looping ones accept more than one. In pattern matching symbols, in the pattern `a b+ c? d` (or `a`, 
+followed by *one or more* `b`'s, optionally followed by a `c`, followed by a `d`), `a`, `c?`, and `d` are 
+singleton patterns, while `b+` is a looping one (see [Quantifiers](#quantifiers)). In addition, each state 
+can have one or more *conditions* based on which it accepts events (see [Conditions](#conditions)).
+
+#### Quantifiers
+
+In FlinkCEP, looping patterns can be specified using the methods: `pattern.oneOrMore()`, for states that expect one or
+more occurrences of a given event (e.g. the `b+` mentioned previously), and `pattern.times(#ofTimes)` for states that 
+expect a specific number of occurrences of a given type of event, e.g. 4 `a`'s. All states, looping or not, can be made 
+optional using the `pattern.optional()` method. For a state named `start`, the following are valid quantifiers:
+ 
+ <div class="codetabs" markdown="1">
+ <div data-lang="java" markdown="1">
+ {% highlight java %}
+ // expecting 4 occurrences
+ start.times(4);
+  
+ // expecting 0 or 4 occurrences
+ start.times(4).optional();
+ 
+ // expecting 1 or more occurrences
+ start.oneOrMore();
+   
+ // expecting 0 or more occurrences
+ start.oneOrMore().optional();
+ {% endhighlight %}
+ </div>
+ 
+ <div data-lang="scala" markdown="1">
+ {% highlight scala %}
+ // expecting 4 occurrences
+ start.times(4)
+   
+ // expecting 0 or 4 occurrences
+ start.times(4).optional()
+  
+ // expecting 1 or more occurrences
+ start.oneOrMore()
+    
+ // expecting 0 or more occurrences
+ start.oneOrMore().optional()
+ {% endhighlight %}
+ </div>
+ </div>
+
+#### Conditions
+
+At every state, and in order to go from one state to the next, you can specify additional **conditions**. 
+These conditions can be related to:
+ 
+ 1. a [property of the incoming event](#conditions-on-properties), e.g. its value should be larger than 5, 
+ or larger than the average value of the previously accepted events.
+
+ 2. the [contiguity of the matching events](#conditions-on-contiguity), e.g. detect pattern `a,b,c` without 
```

Comment:
I would move this to the combining patterns section. It is easier to reason about after understanding the contiguity between single patterns/states.